searchlib/src/vespa/searchlib/queryeval/elementiterator.cpp

```diff
@@ -0,0 +1,71 @@
+#include "elementiterator.h"
+#include <vespa/searchlib/fef/termfieldmatchdata.h>
+#include <vespa/vespalib/objects/objectvisitor.h>
+
+namespace search::queryeval {
+
+void
+ElementIterator::visitMembers(vespalib::ObjectVisitor &visitor) const {
+    visit(visitor, "iterator", _search.get());
+}
+
+ElementIteratorWrapper::ElementIteratorWrapper(SearchIterator::UP search, fef::TermFieldMatchData & tfmd)
+    : ElementIterator(std::move(search)),
+      _tfmd(tfmd)
+{}
+
+ElementIteratorWrapper::~ElementIteratorWrapper() = default;
+
+void
+ElementIteratorWrapper::getElementIds(uint32_t docId, std::vector<uint32_t> & elementIds) {
+    _search->unpack(docId);
+    int prevId(-1);
+    for (auto element : _tfmd) {
+        uint32_t id(element.getElementId());
+        if (prevId != int(id)) {
+            elementIds.push_back(id);
+            prevId = id;
+        }
+    }
+}
+
+void
+ElementIteratorWrapper::mergeElementIds(uint32_t docId, std::vector<uint32_t> & elementIds) {
+    _search->unpack(docId);
+    size_t toKeep(0);
+    int32_t id(-1);
+    auto it = _tfmd.begin();
+    for (int32_t candidate : elementIds) {
+        if (candidate >= id) {
```

Comment:
possible re-write: avoid seeking for the same candidate multiple times

Reply:
Fixed